speechbrain/lobes/models/huggingface/transformers.py

```diff
@@ -0,0 +1,274 @@
+        nested modules of the model are Frozen by invoking: `eval(f"self.model.{nested_freeze_call}()")`
+        Default (None) skips that step.
+    cache_dir: str or Path (default: None)
+        Location of HuggingFace cache for storing pre-trained models, to which symlinks are created.
+
+    Example
+    -------
+    >>> inputs = torch.rand([10, 600])
+    >>> model_hub = "facebook/wav2vec2-base-960h"
+    >>> save_path = "tmp"
+    >>> model = HuggingFaceTransformer(model_hub, save_path=save_path)
+    >>> outputs = model(inputs)
+    """
+
+    def __init__(
+        self,
+        source,
+        save_path,
+        auto_class=AutoModel,
+        for_pretraining_cls=None,
+        forward_partial_fn: Union[Callable, None] = None,
+        modify_state_dict_partial_fn: Union[Callable, None] = None,
+        override_hf_config_partial_fn: Union[Callable, None] = None,
+        override_hf_model_partial_fn: Union[Callable, None] = None,
+        freeze=True,
+        freeze_nested_models_their_calls: Union[List[str], str, None] = None,
+        cache_dir: Union[str, pathlib.Path, None] = "pretrained_models",
+    ):
+        super().__init__()
+
+        # Is the auto_class valid?
+        assert (
+            auto_class in HUGGINGFACE_AUTO_CLASSES
+        ), "Error: please provide a HuggingFace Auto[Class]"
+
+        # Fetch config
+        config, _unused_kwargs = AutoConfig.from_pretrained(
+            source,
+            cache_dir=cache_dir if for_pretraining_cls is None else save_path,
+            return_unused_kwargs=True,
+        )
+
+        # Adjust config as desired
+        if override_hf_config_partial_fn is not None:
+            config = override_hf_config_partial_fn(config)
+
+        # Instantiate model
+        if for_pretraining_cls is not None:
+            # Construct for pretraining
+            self.model = for_pretraining_cls(config)
+        else:
+            # Fetch model architecture
+            if (
+                hasattr(config, "auto_map")
+                and auto_class.__name__ in config.auto_map
+            ):
+                model = auto_class.from_config(config, cache_dir=cache_dir)
+            else:  # AutoModel.from_config case: type(config) in AutoModel._model_mapping.keys() /or: raise ValueError
+                model = auto_class.from_config(config)
+
+            # Download model
+            self._from_pretrained(
+                source,
+                config=config,
+                model=model,
+                modify_state_dict_partial_fn=modify_state_dict_partial_fn,
+                save_path=save_path,
+                cache_dir=cache_dir,
+            )
+
+        # Adjust model as desired
+        if override_hf_model_partial_fn is not None:
+            self.model = override_hf_model_partial_fn(self.model)
+
+        # Set inner forward function
+        if forward_partial_fn is not None:
+            if isinstance(forward_partial_fn, partial):
```

Comment:
Why is this check necessary?

Reply:
good catch!

Reply:
also getting back here, just spotted the reason more clearly.

the outer part
```python
if forward_partial_fn is not None:
   ...
else:
    self.forward_partial_fn = partial(default, model=self.model)
```
is there to put a default forward.

The inner part
```python
  if isinstance(forward_partial_fn, partial):
      self.forward_partial_fn = forward_partial_fn
      self.forward_partial_fn.keywords["model"] = self.model
  else:
      self.forward_partial_fn = partial(
          forward_partial_fn, model=self.model
      )
```
sets the model as a parameter. If the function is a partial, then the model needs to be added to the keywords dict, so when the partial is called later on, this parameter is set correctly. If `forward_partial_fn` is not a partial (but a function), a partial is created.

I put two comments in (will come with the next push).